pallets/builder-incentives/src/lib.rs

```diff
@@ -0,0 +1,315 @@
+			beneficiary: AccountIdOf<T>,
+			// The amount of incentives withdrawn.
+			amount: BalanceOf<T>,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Unable to claim rewards for the current era.
+		CannotClaimRewardsForCurrentEra,
+		/// The smart contract is already registered.
+		ContractAlreadyRegistered,
+		/// The smart contract is not registered.
+		ContractNotRegistered,
+		/// The caller is not the beneficiary authorized to claim rewards.
+		NotAuthorizedToClaimRewards,
+	}
+
+	#[pallet::genesis_config]
+	pub struct GenesisConfig<T: Config> {
+		pub starting_era: EraNumber,
+		#[serde(skip)]
+		pub _config: core::marker::PhantomData<T>,
+	}
+
+	impl<T: Config> Default for GenesisConfig<T> {
+		fn default() -> Self {
+			Self { starting_era: 0, _config: core::marker::PhantomData }
+		}
+	}
+
+	#[pallet::genesis_build]
+	impl<T: Config> BuildGenesisConfig for GenesisConfig<T> {
+		fn build(&self) {
+			let era_duration = T::EraDuration::get();
+			let era_info = EraInfo::<T> {
+				next_era_start: era_duration
+					.saturating_add(BlockNumberFor::<T>::zero())
+					.saturated_into(),
+				contract_fee_total: BalanceOf::<T>::zero(),
+				total_fee_amount: BalanceOf::<T>::zero(),
+			};
+			EraInformation::<T>::insert(self.starting_era, era_info);
+		}
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_initialize(now: BlockNumberFor<T>) -> Weight {
+			let now = now.saturated_into();
+			let mut current_era = CurrentEra::<T>::get();
+			let mut current_era_info = EraInformation::<T>::get(current_era);
+			let mut consumed_weight = T::DbWeight::get().reads(2);
+
+			// Check if is new era, if not don't do anything.
+			if !current_era_info.is_new_era(now) {
+				return consumed_weight;
+			}
+			// If is new era, update the new era information.
+			current_era.saturating_inc();
+			current_era_info
+				.set_next_era_start(now.saturating_add(T::EraDuration::get().saturated_into()));
+			current_era_info.reset_contract_fee();
+			current_era_info.resent_total_fee();
+			// Update storage items
+			CurrentEra::<T>::put(current_era);
+			EraInformation::<T>::insert(current_era, current_era_info);
+			consumed_weight = T::DbWeight::get().writes(2);
+			Self::deposit_event(Event::<T>::NewEra { era: current_era });
+			consumed_weight
+		}
+		// TODO: Implement on_idle for handling rewards not claimed after X period
+		// fn on_idle(_block: BlockNumberFor<T>, remaining_weight: Weight) -> Weight {
+		// }
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Register a new contract for builder incentives.
+		///
+		/// # Parameters
+		/// - `beneficiary`: The account that will be the beneficiary of the contract incentives.
+		/// - `contract`: The smart contract's account to be registered.
+		#[pallet::call_index(0)]
+		#[pallet::weight(Weight::zero())]
+		pub fn register_contract(
+			origin: OriginFor<T>,
+			beneficiary: T::AccountId,
+			contract: AccountContractId,
```

Comment:
This needs to be changed. The contract should be the origin, removing the need for this parameter.